extensions/vertx-http/deployment/src/test/java/io/quarkus/vertx/http/cors/CORSSecurityTestCase.java

```diff
@@ -98,59 +98,46 @@ public void corsPreflightTest() {
                 .when()
                 .auth().basic("user", "user")
                 .options("/test").then()
-                .statusCode(403)
+                .statusCode(200)
```

Comment:
@sberyozkin @stuartwdouglas Yeah, when the cors preflight response statuscode is 200 but `Access-Control-Allow-Methods` does not contain the method which is inside `Access-Control-Request-Method` the browser was not blocking the actual request. If you like you can check it for yourself here https://github.com/hamburml/quarkus-cors-example.

Reply:
This is a pre-flight options request. It is not the actual request. The important part is the value of the Access-Control-Allowed-Methods etc, not the return code.

If the response code was important then none of these response headers would be required, the server would just return a 200 or a 403 based on if it is allowed or not.

If you look at the fetch spec in particular this is called out here: https://fetch.spec.whatwg.org/#http-responses

`They can provide a static response. This can be helpful when working with caching intermediaries. A static response can both be successful and not successful depending on the [CORS request](https://fetch.spec.whatwg.org/#cors-request). This is okay.`

If the user has specified all the different properties in our CORS config we could just return a static response every time. Our response is dynamic because if the user has not specified a value we respond with what is requested in the HTTP request. 

We still return a 403 to the actual POST request. CORS will never block POST,GET or HEAD no matter the value of Access-Control-Allowed-Methods (basically because the JS can just create a form an submit the form and get a POST request that way). If you change your example to PUT rather than post and use this version of Quarkus you will see the browser blocking it.

Reply:
Yes, I know that it is the preflight options request and not the actual request.

> CORS will never block POST,GET or HEAD no matter the value of Access-Control-Allowed-Methods

That is true. Cors it not a firewall so it can not block the request. But the browser can decide to not carry out the actual request and it uses the response of the cors preflight as decisive factors. And in my example this only worked when the cors preflight options request had a 403 status code. 

> just return a static response every time

Nah, in our aws api gateway we had a static response for cors options request. But the `Access-Control-Allow-Origin` should always be the origin of the request if you have more than one origin who is allowed to use this resouce.

> If you change your example to PUT rather than post and use this version of Quarkus you will see the browser blocking it.

Thanks, Ill try it out later. I simply do not understand why CORS never blocks POST,GET,HEAD if I can remove them from the allowed-methods list. It would be unambiguous to not allow POST,GET,HEAD in Access-Control-Allowed-Methods because it simply does not matter (because never blocked). Thx for your explanation.

Reply:
It is because there are workarounds to invoking a POST/GET request. To do a GET request the JS can just create an <image> element with the source as the GET request to invoke. To do a POST request the JS can just create a FORM element and then call form.submit().

Blocking these requests would just create the illusion of security, as the JS could still invoke them.

Reply:
oh wow, now i got it. Thx!

Reply:
@stuartwdouglas Sure, but returning `403` would be consistent with how we handle mismatched Origins for the preflights. This filter did return `200` for mismatched origins for preflight requests but then we changed it to `403` after some discussions which you may recall (incl for it to be consistent with Vert.x)
 
I see, if we block the method for an actual POST/etc request anyway, then, for the preflight request, it does not really matter if it is 403 or not, but if 403 can prevent some browsers even trying to follow up with an actual request then it should be good.
 
Can we agree on `403` in case of preflight method or header mis-matches ?

Reply:
Header mismatch is an absolute no-go, it's not how headers work. Access-Control-Allow-Headers tells the browser the headers that the JS can set on the request. The browser can and does add its own headers outside this set. 

Methods are also bad IMHO. If we return a 200 the browser can cache this information and use it to avoid further pre-flight requests. I think even returning a 403 on origin mismatch for preflight is potentially wrong, however the cache is not really an issue there so it is probably fine.

Like I said earlier the browser blocking on a 403 for incorrect method only adds the illusion of security.

Reply:
@stuartwdouglas Makes sense now, thanks for the clarifications about the cache, sounds good